kolibri/plugins/document_pdf_render/assets/src/views/index.vue

```diff
@@ -175,49 +219,69 @@
       pageRef(index) {
         return `pdfPage-${index}`;
       },
-      checkPages() {
+      // debouncing so we're not de/re-render many pages unnecessarily
+      checkPages: debounce(function() {
         const top = this.$refs.pdfcontainer.scrollTop;
         const bottom = top + this.$refs.pdfcontainer.clientHeight;
         const topPageNum = Math.ceil(top / this.pageHeight);
         const bottomPageNum = Math.ceil(bottom / this.pageHeight);
-        let i;
-        // Loop through all pages, show ones that are in the display window,
-        // hide ones that are not
-        for (i = 1; i <= this.totalPages; i++) {
+
+        // Loop through all pages, show ones that are in the display window, hide ones that aren't
+        for (let i = 1; i <= this.totalPages; i++) {
           if (i < topPageNum - pageDisplayWindow || i > bottomPageNum + pageDisplayWindow) {
             this.hidePage(i);
           } else {
             this.showPage(i);
           }
         }
-      },
+      }, renderDebounceTime),
     },
     watch: {
       scrollPos: 'checkPages',
+      scale(newScale, oldScale) {
+        const noChange = newScale === oldScale;
+        const firstChange = oldScale === null;
+        if (!noChange && !firstChange) {
+          Object.keys(this.pdfPages).forEach(pageNum => {
+            // toggle between hide and show to re-render the page
+            this.hidePage(Number(pageNum));
+            this.showPage(Number(pageNum));
+          });
+          this.checkPages();
+        }
+        this.checkPages();
+      },
     },
     created() {
-      this.pdfPages = {};
-      this.pdfloadingPromise = PDFJSLib.getDocument(
-        this.defaultFile.storage_url,
-        null,
-        null,
-        progress => {
-          this.progress = progress.loaded / progress.total;
-        }
-      ).then(pdfDocument => {
-        this.loading = false;
-        this.totalPages = pdfDocument.numPages;
-        // Track the pdf document
-        this.pdfDocument = pdfDocument;
-        // Begin retrieving the first page
-        return this.getPage(1);
-      });
+      this.loadPdfPromise = PDFJSLib.getDocument(this.defaultFile.storage_url);
     },
     mounted() {
-      this.pdfloadingPromise.then(() => {
-        this.showPage(1);
-        this.$emit('startTracking');
+      if (this.fullscreenAllowed) {
+        ScreenFull.onchange(() => {
+          this.isFullscreen = ScreenFull.isFullscreen;
+        });
+      }
+
+      // pass callback to update loading bar
+      this.loadPdfPromise.onProgress = loadingProgress => {
+        this.progress = loadingProgress.loaded / loadingProgress.total;
+      };
+
+      // Retrieve the document and its corresponding object
+      this.loadPdfPromise.then(pdfDocument => {
+        this.pdfDocument = pdfDocument;
+        this.totalPages = pdfDocument.numPages;
+        this.pdfPages = {};
+
+        // Retrieve the first Page object
+        this.getPage(1).then(pdfPage => {
```

Comment:
Note that you are now deferring getting the first page until the component has mounted - this should be moved back into the `created` handler, so that we are not waiting on DOM mounting for this operation. The only thing we need to wait for DOM mounting for is actually rendering into the page.